pkg/infrastructure/clusterapi/clusterapi.go

```diff
@@ -0,0 +1,275 @@
+package clusterapi
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"gopkg.in/yaml.v2"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	utilkubeconfig "sigs.k8s.io/cluster-api/util/kubeconfig"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/manager/signals"
+
+	"github.com/openshift/installer/pkg/asset"
+	"github.com/openshift/installer/pkg/asset/cluster/metadata"
+	"github.com/openshift/installer/pkg/asset/ignition/bootstrap"
+	"github.com/openshift/installer/pkg/asset/ignition/machine"
+	"github.com/openshift/installer/pkg/asset/installconfig"
+	"github.com/openshift/installer/pkg/asset/kubeconfig"
+	"github.com/openshift/installer/pkg/asset/manifests/capiutils"
+	capimanifests "github.com/openshift/installer/pkg/asset/manifests/clusterapi"
+	"github.com/openshift/installer/pkg/clusterapi"
+	"github.com/openshift/installer/pkg/infrastructure"
+	"github.com/openshift/installer/pkg/types"
+)
+
+// InfraProvider is the base implementation for provisioning cluster
+// infrastructure using CAPI. Platforms should embed this struct and
+// implement:
+// .
+type InfraProvider struct {
+	CAPIInfraHelper
+}
+
+func InitializeProvider(platform CAPIInfraHelper) infrastructure.Provider {
+	return InfraProvider{platform}
+}
+
+type CAPIInfraHelper interface {
+	// PreProvision is called before provisioning using CAPI controllers has begun.
+	// and should be used to create dependencies needed for CAPI provisioning,
+	// such as IAM roles or policies.
+	PreProvision(in PreProvisionInput) error
+
+	// ControlPlaneAvailable is called once cluster.Spec.ControlPlaneEndpoint.IsValid()
+	// returns true, typically after load balancers have been provisioned. It can be used
+	// to create DNS records.
+	ControlPlaneAvailable(in ControlPlaneAvailableInput) error
+}
+
+type PreProvisionInput struct {
+	ClusterID     string
+	InstallConfig *installconfig.InstallConfig
+}
+
+type ControlPlaneAvailableInput struct {
+	Cluster       *clusterv1.Cluster
+	InstallConfig *installconfig.InstallConfig
+}
```

Comment:
What about adding the `client.Client` here? 

In #7879 (which is the aws implementation of this), that is how I'm grabbing the subnets and machine data.

Reply:
sgtm, either we pass it or we can grab it from the global system; might be better for long term usage to pass it, as we can also unit test it that way